pkg/storage/db.go

```diff
@@ -0,0 +1,283 @@
+package storage
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/asecurityteam/asset-inventory-api/pkg/domain"
+	"github.com/asecurityteam/asset-inventory-api/pkg/logs"
+	"github.com/asecurityteam/runhttp"
+	_ "github.com/lib/pq" // postgres driver for sql must be imported so sql finds and uses it
+)
+
+const tableAWSResources = "aws_resources"
+const tableAWSIPS = "aws_ips"
+const tableAWSHostnames = "aws_hostnames"
+const tableAWSEventsIPSHostnames = "aws_events_ips_hostnames"
+
+const added = "ADDED" // one of the network event types we track
+
+// can't use Sprintf in a const, so...
+const commonQuery = "SELECT " +
+	"	aws_resources_id, aws_ips_ip, aws_hostnames_hostname, is_public, is_join, ts, aws_resources.account_id, aws_resources.region, aws_resources.type, aws_resources.meta " +
+	"FROM " +
+	"	" + tableAWSEventsIPSHostnames + " " +
+	"JOIN " + tableAWSResources + " " +
+	"	ON aws_resources_id = aws_resources.id " +
+	"WHERE "
+
+// DB represents a convenient database abstraction layer
+type DB struct {
+	sqldb *sql.DB // this is a unit test seam
+	once  sync.Once
+}
+
+// Init initializes a connection to a Postgres database according to the environment variables POSTGRES_HOST, POSTGRES_PORT, POSTGRES_USER, POSTGRES_PASSWORD, POSTGRES_DATABASE
+func (db *DB) Init(ctx context.Context, postgresConfig *PostgresConfig) error {
+	var initerr error
+	db.once.Do(func() {
+
+		host := postgresConfig.Hostname
+		port := postgresConfig.Port
+		user := postgresConfig.Username
+		password := postgresConfig.Password
+		dbname := postgresConfig.DatabaseName
+
+		if db.sqldb == nil {
+			sslmode := "disable"
+			if host != "localhost" {
+				sslmode = "enable"
+			}
+			psqlInfo := fmt.Sprintf("host=%s port=%s user=%s "+
+				"password=%s dbname=%s sslmode=%s",
+				host, port, user, password, dbname, sslmode)
+			pgdb, err := sql.Open("postgres", psqlInfo)
+			if err != nil {
+				initerr = err
+			}
+
+			err = pgdb.Ping()
+			if err != nil {
+				initerr = err
+			}
+
+			db.sqldb = pgdb
+		}
+	})
+	return initerr
+}
+
+// StoreCloudAsset an implementation of the Storage interface that records to a database
+func (db *DB) StoreCloudAsset(ctx context.Context, cloudAssetChanges domain.CloudAssetChanges) error {
+
+	logger := runhttp.LoggerFromContext(ctx)
+
+	tx, err := db.sqldb.Begin()
+	if err != nil {
+		return err
+	}
+
+	defer func() {
```

Comment:
This `defer` is guaranteed to run before the function returns a value. We've never used it as most cases end up seeming like anti-patterns but there is the option of defining the method with a named return value. For example: https://play.golang.org/p/9ScRVbnxw_W .

I'd be happy if we never used named return values but SQL style transactions may be one place where they fit. Normally I'd rather have a log in non-handler code than named returns but the current state of this method is that it could return `nil` but actually fail.

I'm OK starting with this so long as we track it and figure out what we want to do long term before going to production.

Reply:
I'm gonna do what Zach suggested in the next comment.